src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+                {
+                    CleanupStreamAndWatcher();
+                    CreateStreamAndStartWatcher();
+                    _lastWatchedDirectory = _directory;
+                }
+            }
+        }
+
+        private void CreateStreamAndStartWatcher()
+        {
+            Debug.Assert(_eventStream == IntPtr.Zero);
+            Debug.Assert(_watcherRunLoop == IntPtr.Zero);
+
+            // Make sure we only do this if there is a valid directory
+            if (String.IsNullOrEmpty(_directory) == false)
+            {
+                // Get the path to watch and put it into a CFArray.
+                CFStringRef path = Interop.CoreFoundation.CFStringCreateWithCString(IntPtr.Zero, _directory, Interop.CoreFoundation.CFStringBuiltInEncodings.kCFStringEncodingUTF8);
+                CFArrayRef arrPaths = Interop.CoreFoundation.CFArrayCreate(IntPtr.Zero, new CFStringRef[1] { path }, 1, IntPtr.Zero);
+
+                // Create the event stream for the path and tell the stream to watch for file system events.
+                _eventStream = Interop.CoreServices.FSEventStreamCreate(
+                    IntPtr.Zero,
+                    FileSystemEventCallback,
+                    IntPtr.Zero,
+                    arrPaths,
+                    Interop.CoreServices.kFSEventStreamEventIdSinceNow,
+                    0.2f,
+                    EventStreamFlags);
+                Debug.Assert(_eventStream != IntPtr.Zero);
+
+                // The paths are copied inside FSEventStreamCreate so delete our references
+                Interop.CoreFoundation.CFRelease(path);
+                Interop.CoreFoundation.CFRelease(arrPaths);
+
+                // Create and start our watcher thread then wait for the thread to initialize and start 
+                // the RunLoop. We wait for that to prevent this function from returning before the RunLoop
+                // has a chance to start so that any callers won't race with the background thread's initialization
+                // and calling Stop, which would attempt to stop a RunLoop that hasn't started yet.
+                _watcherThread = new Thread(new ThreadStart(WatchForFileSystemEventsThreadStart));
+                _watcherThread.Start();
+                _runLoopStartedEvent.WaitOne();
+            }
+        }
+
+        private void CleanupStreamAndWatcher()
+        {
+            // Make sure there's a loop to clear
+            if (_watcherRunLoop != IntPtr.Zero)
+            {
+                // Stop the RunLoop and wait for the thread to exit gracefully
+                Interop.CoreFoundation.CFRunLoopStop(_watcherRunLoop);
+                _watcherThread.Join();
+            }
+
+            // Clean up the EventStream, if it exists
+            if (_eventStream != IntPtr.Zero)
+            {
+                Interop.CoreFoundation.CFRelease(_eventStream);
+            }
+
+            // Reset back to zero
+            _eventStream = IntPtr.Zero;
```

Comment:
You could move these resets back up into the if blocks... no need to set them to zero if they already are.